firebase-firestore/src/androidTest/java/com/google/firebase/firestore/CountTest.java

```diff
@@ -0,0 +1,249 @@
+    AggregateQuery query2_same =
+        coll1.document("bar").collection("baz").whereEqualTo("a", 1).limit(100).count();
+    AggregateQuery query3 =
+        coll1.document("bar").collection("baz").whereEqualTo("b", 1).orderBy("c").count();
+    AggregateQuery query3_same =
+        coll1.document("bar").collection("baz").whereEqualTo("b", 1).orderBy("c").count();
+
+    assertEquals(query1, query1_same);
+    assertEquals(query2, query2_same);
+    assertEquals(query3, query3_same);
+
+    assertEquals(query1.hashCode(), query1_same.hashCode());
+    assertEquals(query2.hashCode(), query2_same.hashCode());
+    assertEquals(query3.hashCode(), query3_same.hashCode());
+
+    assertNotEquals(null, query1);
+    assertNotEquals(query1, query2);
+    assertNotEquals(query2, query3);
+    assertNotEquals(query1.hashCode(), query2.hashCode());
+    assertNotEquals(query2.hashCode(), query3.hashCode());
+  }
+
+  @Test
+  public void testCanRunCount() {
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "a", map("k", "a"),
+                "b", map("k", "b"),
+                "c", map("k", "c")));
+
+    AggregateQuerySnapshot snapshot =
+        waitFor(collection.count().get(AggregateSource.SERVER_DIRECT));
+    assertEquals(Long.valueOf(3), snapshot.getCount());
+  }
+
+  @Test
+  public void testCanRunCountWithFilters() {
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "a", map("k", "a"),
+                "b", map("k", "b"),
+                "c", map("k", "c")));
+
+    AggregateQuerySnapshot snapshot =
+        waitFor(collection.whereEqualTo("k", "b").count().get(AggregateSource.SERVER_DIRECT));
+    assertEquals(Long.valueOf(1), snapshot.getCount());
+  }
+
+  @Test
+  public void testSnapshotEquals() {
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "a", map("k", "a"),
+                "b", map("k", "b"),
+                "c", map("k", "c")));
+
+    AggregateQuerySnapshot snapshot1 =
+        waitFor(collection.whereEqualTo("k", "b").count().get(AggregateSource.SERVER_DIRECT));
+    AggregateQuerySnapshot snapshot1_same =
+        waitFor(collection.whereEqualTo("k", "b").count().get(AggregateSource.SERVER_DIRECT));
+
+    AggregateQuerySnapshot snapshot2 =
+        waitFor(collection.whereEqualTo("k", "a").count().get(AggregateSource.SERVER_DIRECT));
+    waitFor(collection.document("d").set(map("k", "a")));
+    AggregateQuerySnapshot snapshot2_different =
+        waitFor(collection.whereEqualTo("k", "a").count().get(AggregateSource.SERVER_DIRECT));
+
+    assertEquals(snapshot1, snapshot1_same);
+    assertEquals(snapshot1.hashCode(), snapshot1_same.hashCode());
+    assertEquals(snapshot1.getQuery(), collection.whereEqualTo("k", "b").count());
+
+    assertNotEquals(snapshot1, snapshot2);
```

Comment:
Add tests for invoking AggregateQuerySnapshot.equals() with 
1. null.
2. a non-null, non-AggregateQuerySnapshot object (e.g. a String).